lib/runtime/storage/storagediff.go

```diff
@@ -0,0 +1,308 @@
+		upserts:        make(map[string][]byte),
+		deletes:        make(map[string]bool),
+		childChangeSet: make(map[string]*storageDiff),
+		mtx:            sync.RWMutex{},
+	}
+}
+
+// get retrieves the value associated with the key if it's present in the
+// change set and returns a boolean indicating if the key is marked for deletion
+func (cs *storageDiff) get(key string) ([]byte, bool) {
+	if cs == nil {
+		return nil, false
+	}
+
+	cs.mtx.RLock()
+	defer cs.mtx.RUnlock()
+
+	// Check in recent upserts if not found check if we want to delete it
+	if val, ok := cs.upserts[key]; ok {
+		return val, false
+	} else if deleted := cs.deletes[key]; deleted {
+		return nil, true
+	}
+
+	return nil, false
+}
+
+// upsert records a new value for the key, or updates an existing value.
+// If the key was previously marked for deletion, that deletion is undone
+func (cs *storageDiff) upsert(key string, value []byte) {
+	if cs == nil {
+		return
+	}
+
+	cs.mtx.Lock()
+	defer cs.mtx.Unlock()
+	// If we previously deleted this trie we have to undo that deletion
+	if cs.deletes[key] {
+		delete(cs.deletes, key)
+	}
+
+	cs.upserts[key] = value
+}
+
+// delete marks a key for deletion and removes it from upserts and
+// child changesets, if present.
+func (cs *storageDiff) delete(key string) {
+	if cs == nil {
+		return
+	}
+
+	cs.mtx.Lock()
+	defer cs.mtx.Unlock()
+
+	delete(cs.childChangeSet, key)
+	delete(cs.upserts, key)
+	cs.deletes[key] = true
+}
+
+// deleteChildLimit deletes lexicographical sorted keys from a child trie with
+// a maximum limit, potentially marking the entire child trie for deletion
+// if the limit is exceeded.
+// This method will do not consider keys created during actual block execution
+// https://spec.polkadot.network/chap-host-api#id-version-2-prototype-2
+func (cs *storageDiff) deleteChildLimit(keyToChild string,
+	currentChildKeys []string, limit int) (
+	deleted uint32, allDeleted bool) {
+	childChanges := cs.childChangeSet[keyToChild]
+	if childChanges == nil {
+		childChanges = newStorageDiff()
+	}
+
+	if limit == -1 {
+		cs.delete(keyToChild)
+		deletedKeys := len(childChanges.upserts) + len(currentChildKeys)
+		return uint32(deletedKeys), true
+	}
+
+	newKeys := maps.Keys(childChanges.upserts)
+	allKeys := append(newKeys, currentChildKeys...)
+	sort.Strings(allKeys)
+
+	for _, k := range allKeys {
+		if limit == 0 {
+			break
+		}
+		childChanges.delete(k)
+		deleted++
+		// Do not consider keys created during actual block execution
+		if !slices.Contains(newKeys, k) {
+			limit--
+		}
+	}
+	cs.childChangeSet[keyToChild] = childChanges
+
+	return deleted, deleted == uint32(len(allKeys))
+}
+
+// clearPrefixInChild clears keys with a specific prefix within a child trie.
+func (cs *storageDiff) clearPrefixInChild(keyToChild string, prefix []byte,
+	childKeys []string, limit int) (deleted uint32, allDeleted bool) {
+	childChanges := cs.childChangeSet[keyToChild]
+	if childChanges == nil {
+		childChanges = newStorageDiff()
+	}
```

Comment:
if the child does not exist then we can just return w/o any other operation, right?

Reply:
In this scenario, the absence of the child in the diff indicates no updates were made to its keys/values during this block's execution. However, the child may still exist in the current trie. Therefore, we need to generate a `newStorageDiff` to clear any keys prefixed from the current state child